<h1>Simplilearn Cat and Dog Classification</h1>
<h3>Submitted by Aditya Retissin Poozhiyil</h3>

<h3><u>Project Description</u></h3>

Build a CNN model that classifies the given pet images correctly into dog and cat images. 

The project scope document specifies the requirements for the project “Pet Classification Model Using CNN.” Apart from specifying the functional and nonfunctional requirements for the project, it also serves as an input for project scoping. 

Project Description and Scope 

You are provided with the following resources that can be used as inputs for your model: 

1. A collection of images of pets, that is, cats and dogs. These images are of 

different sizes with varied lighting conditions. 2. Code template containing the following code blocks: 

a. Import modules (part 1) b. Set hyper parameters (part 2) c. Read image data set (part 3) d. Run TensorFlow model (part 4) 

You are expected to write the code for CNN image classification model (between Parts 3 and 4) using TensorFlow that trains on the data and calculates the accuracy score on the test data. 

Project Guidelines 

Begin by extracting ipynb file and the data in the same folder. The CNN model (cnn_model_fn) should have the following layers: 

● Input layer 

● Convolutional layer 1 with 32 filters of kernel size[5,5] 

● Pooling layer 1 with pool size[2,2] and stride 2 

● Convolutional layer 2 with 64 filters of kernel size[5,5] 

● Pooling layer 2 with pool size[2,2] and stride 2 

● Dense layer whose output size is fixed in the hyper parameter: fc_size=32 

● Dropout layer with dropout probability 0.4 

Predict the class by doing a softmax on the output of the dropout layers. 

This should be followed by training and evaluation: 

● For the training step, define the loss function and minimize it 

● For the evaluation step, calculate the accuracy 

Run the program for 100, 200, and 300 iterations, respectively. Follow this by a report on the final accuracy and loss on the evaluation data. 

In [239]:
pip install notebook-as-pdf

     |████████████████████████████████| 153kB 2.8MB/s 
     |████████████████████████████████| 10.3MB 7.4MB/s 
     |████████████████████████████████| 133kB 46.7MB/s 
     |████████████████████████████████| 81kB 9.6MB/s 
     |████████████████████████████████| 71kB 8.6MB/s 
ERROR: kaggle 1.5.8 has requirement urllib3<1.25,>=1.21.1, but you'll have urllib3 1.25.10 which is incompatible.
ERROR: datascience 0.10.6 has requirement folium==0.2.1, but you'll have folium 0.8.3 which is incompatible.
  Found existing installation: urllib3 1.24.3
    Uninstalling urllib3-1.24.3:
      Successfully uninstalled urllib3-1.24.3
  Found existing installation: tqdm 4.41.1
    Uninstalling tqdm-4.41.1:
      Successfully uninstalled tqdm-4.41.1


In [ ]:
import os
import numpy as np

In [2]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [3]:
os.chdir("/content/drive/My Drive/Python/Practice/Cat vs Dogs")

In [4]:
ls

'Cat Vs Dogs.ipynb'   test/   train/   train.npy


<h2>Importing the required library's</h2>

In [207]:
import seaborn as sns
import cv2
from random import shuffle
from sklearn.preprocessing import LabelEncoder
from keras.utils import to_categorical
import matplotlib.pyplot as plt

In [208]:
TRAIN_PATH = "train/"
TEST_PATH = "test/"
IMG_SIZE = (100,100)

*   Iterating through each class and assigning label's accordingly 1 for dog , 0 for cat
*   Reading the image and resizing to the img size
*   Appending the label and training data to the corresponding variables




In [209]:
def process_img():
  training_data =[]
  label_data = []
  for classes in os.listdir(TRAIN_PATH):
    path = os.path.join(TRAIN_PATH,classes)
    for img in os.listdir(path):
      label = 1 if classes=="dogs" else 0
      new_img = cv2.imread(os.path.join(path,img))
      new_img = cv2.resize(new_img , IMG_SIZE)
      training_data.append(np.array(new_img))
      label_data.append(label)
  return np.array(training_data),label_data

In [210]:
X , Y = process_img()

In [211]:
X.shape

(40, 100, 100, 3)

Converting the label's to categorical format so it can be fed to the convolutional neural net

In [212]:
Y = to_categorical(Y,2)

<h2>Model building</h2>

<h3>Importing the required library's</h3>

In [213]:
from tensorflow.keras.layers import Dropout
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense
from tensorflow.keras.layers import Activation
from tensorflow.keras.layers import Conv2D
from tensorflow.keras.layers import MaxPooling2D
from tensorflow.keras.layers import Flatten
from keras.optimizers import Adam
from tensorflow.keras.layers import Softmax

*   Training the model with the learning rate of 1e-5
*   And Optimizer : Adam



In [214]:
INPUT_SHAPE = (100,100,3)
learning_rate = 1e-5
optimizer = Adam(lr=learning_rate)

In [215]:
model = Sequential()

model.add(Conv2D(32, (5, 5), input_shape=(INPUT_SHAPE),padding='same'))
model.add(Activation("relu"))
model.add(MaxPooling2D(pool_size=(2, 2)))

model.add(Conv2D(64,(5,5),padding='same'))
model.add(Activation("relu"))
model.add(MaxPooling2D(pool_size=(2, 2),strides=(2,2)))

model.add(Flatten())
model.add(Dense(32))
model.add(Dropout(0.4))
model.add(Dense(2))
model.add(Activation("softmax"))

In [216]:
model.compile(loss="categorical_crossentropy",
              optimizer=optimizer,
              metrics=["accuracy"])

<h3>Summary of the model</h3>

In [217]:
model.summary()

Model: "sequential_10"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
conv2d_18 (Conv2D)           (None, 100, 100, 32)      2432      
_________________________________________________________________
activation_24 (Activation)   (None, 100, 100, 32)      0         
_________________________________________________________________
max_pooling2d_16 (MaxPooling (None, 50, 50, 32)        0         
_________________________________________________________________
conv2d_19 (Conv2D)           (None, 50, 50, 64)        51264     
_________________________________________________________________
activation_25 (Activation)   (None, 50, 50, 64)        0         
_________________________________________________________________
max_pooling2d_17 (MaxPooling (None, 25, 25, 64)        0         
_________________________________________________________________
flatten_8 (Flatten)          (None, 40000)           

For evaluation purposes we will import confusion matrix as well as accuracy_score from sklearn

In [218]:
from sklearn.metrics import confusion_matrix
from sklearn.metrics import accuracy_score

<h3>Processing test data</h3>

In [222]:
def process_test_data():
  test_data = []
  true_labels = []
  for classes in os.listdir(TEST_PATH):
    path = os.path.join(TEST_PATH,classes)
    for files in os.listdir(path):
      labels = 1 if classes=="dogs" else 0
      new_img = cv2.imread(os.path.join(path,files))
      new_img = cv2.resize(new_img,(IMG_SIZE))
      test_data.append(np.array(new_img))
      true_labels.append(labels)
  return np.array(test_data),true_labels

In [223]:
TEST_INPUT_SHAPE = (-1,100,100,3)

In [224]:
test , true_label = process_test_data()

In [225]:
test_data = test.reshape(TEST_INPUT_SHAPE)

<h2>Training the model</h2>

<h3>For 100 Epochs</h3>

In [226]:
model.fit(X,Y,epochs=100,steps_per_epoch=X.shape[0])

Epoch 1/100
40/40 [==============================] - 1s 37ms/step - loss: 1.9827 - accuracy: 0.8250
Epoch 2/100
40/40 [==============================] - 1s 37ms/step - loss: 0.7120 - accuracy: 0.9500
Epoch 3/100
40/40 [==============================] - 1s 37ms/step - loss: 0.2004 - accuracy: 0.9500
Epoch 4/100
40/40 [==============================] - 1s 37ms/step - loss: 0.6524 - accuracy: 0.9250
Epoch 5/100
40/40 [==============================] - 1s 37ms/step - loss: 0.4392 - accuracy: 0.9750
Epoch 6/100
40/40 [==============================] - 1s 37ms/step - loss: 1.5696 - accuracy: 0.8500
Epoch 7/100
40/40 [==============================] - 1s 37ms/step - loss: 1.0031 - accuracy: 0.9000
Epoch 8/100
40/40 [==============================] - 2s 38ms/step - loss: 0.4351 - accuracy: 0.9500
Epoch 9/100
40/40 [==============================] - 2s 38ms/step - loss: 1.9999 - accuracy: 0.8750
Epoch 10/100
40/40 [==============================] - 2s 38ms/step - loss: 0.1220 - accuracy: 0.9750

In [227]:
prediction = [np.argmax(model.predict(img.reshape(TEST_INPUT_SHAPE))) for img in test_data]

In [228]:
print(confusion_matrix(true_label,prediction))

[[5 5]
 [4 6]]


In [229]:
print (f"The accuracy is : {accuracy_score(true_label,prediction)*100}%")

The accuracy is : 55.00000000000001%


<h3>For 200 Epochs</h3>

In [230]:
model.fit(X,Y,epochs=200,steps_per_epoch=X.shape[0])

Epoch 1/200
40/40 [==============================] - 1s 37ms/step - loss: 1.3479e-04 - accuracy: 1.0000
Epoch 2/200
40/40 [==============================] - 1s 37ms/step - loss: 1.0267e-04 - accuracy: 1.0000
Epoch 3/200
40/40 [==============================] - 1s 37ms/step - loss: 0.5689 - accuracy: 0.9500
Epoch 4/200
40/40 [==============================] - 2s 38ms/step - loss: 0.0000e+00 - accuracy: 1.0000
Epoch 5/200
40/40 [==============================] - 1s 37ms/step - loss: 0.0000e+00 - accuracy: 1.0000
Epoch 6/200
40/40 [==============================] - 1s 37ms/step - loss: 0.0000e+00 - accuracy: 1.0000
Epoch 7/200
40/40 [==============================] - 2s 38ms/step - loss: 0.0487 - accuracy: 0.9750
Epoch 8/200
40/40 [==============================] - 2s 38ms/step - loss: 0.0052 - accuracy: 1.0000
Epoch 9/200
40/40 [==============================] - 1s 37ms/step - loss: 0.3387 - accuracy: 0.9750
Epoch 10/200
40/40 [==============================] - 1s 37ms/step - loss: 1.361

In [231]:
prediction = [np.argmax(model.predict(img.reshape(TEST_INPUT_SHAPE))) for img in test_data]

<h3>Confusion Matrix</h3>

In [232]:
print(confusion_matrix(true_label,prediction))

[[8 2]
 [5 5]]


<h3>Accuracy</h3>

In [233]:
print (f"The accuracy is : {accuracy_score(true_label,prediction)*100}%")

The accuracy is : 65.0%


<h3>For 300 Epochs</h3>

In [234]:
model.fit(X,Y,epochs=300,steps_per_epoch=X.shape[0])

Epoch 1/300
40/40 [==============================] - 1s 37ms/step - loss: 0.0000e+00 - accuracy: 1.0000
Epoch 2/300
40/40 [==============================] - 1s 37ms/step - loss: 0.0000e+00 - accuracy: 1.0000
Epoch 3/300
40/40 [==============================] - 1s 37ms/step - loss: 0.0000e+00 - accuracy: 1.0000
Epoch 4/300
40/40 [==============================] - 2s 38ms/step - loss: 0.0000e+00 - accuracy: 1.0000
Epoch 5/300
40/40 [==============================] - 2s 38ms/step - loss: 0.0000e+00 - accuracy: 1.0000
Epoch 6/300
40/40 [==============================] - 1s 37ms/step - loss: 0.0000e+00 - accuracy: 1.0000
Epoch 7/300
40/40 [==============================] - 1s 37ms/step - loss: 1.4901e-08 - accuracy: 1.0000
Epoch 8/300
40/40 [==============================] - 1s 37ms/step - loss: 0.0000e+00 - accuracy: 1.0000
Epoch 9/300
40/40 [==============================] - 2s 38ms/step - loss: 0.0000e+00 - accuracy: 1.0000
Epoch 10/300
40/40 [==============================] - 1s 37ms/st

In [235]:
prediction = [np.argmax(model.predict(img.reshape(TEST_INPUT_SHAPE))) for img in test_data]

<h3>Confusion Matrix</h3>

In [236]:
print(confusion_matrix(true_label,prediction))

[[5 5]
 [3 7]]


<h3>Accuracy Score</h3>

In [237]:
print (f"The accuracy is : {accuracy_score(true_label,prediction)*100}%")

The accuracy is : 60.0%


We can see that the model with that trained over 200 epochs performed the best with 65% accuracy. The model could improved by feeding in more data